src/transforms/aws_ec2_metadata.rs

```diff
@@ -0,0 +1,596 @@
+}
+
+#[derive(Debug, Deserialize)]
+#[serde(rename_all = "camelCase")]
+struct DynamicIdentityDocument {
+    account_id: String,
+    architecture: String,
+    image_id: String,
+    instance_id: String,
+    instance_type: String,
+    private_ip: String,
+    region: String,
+    version: String,
+}
+
+impl MetadataClient {
+    pub fn new(
+        host: Uri,
+        keys: Keys,
+        state: WriteHandle,
+        refresh_interval: Duration,
+        fields: Vec<Atom>,
+    ) -> Self {
+        Self {
+            client: Client::new(),
+            host,
+            token: None,
+            keys,
+            state,
+            refresh_interval,
+            fields: fields.into_iter().collect(),
+        }
+    }
+
+    async fn fetch_metadata(&mut self) -> Result<(), crate::Error> {
+        loop {
+            self.fetch_all().await?;
+
+            // Make changes viewable to the transform. This may block if
+            // readers are still reading.
+            self.state.refresh();
+
+            let deadline = Instant::now() + self.refresh_interval;
+
+            Delay::new(deadline).compat().await?;
+        }
+    }
+
+    pub async fn get_token(&mut self) -> Result<Bytes, crate::Error> {
+        if let Some((token, next_refresh)) = self.token.clone() {
+            // If the next refresh is greater (in the future) than
+            // the current time we can return the token since its still valid
+            // otherwise lets refresh it.
+            if next_refresh > Instant::now() {
+                return Ok(token);
+            }
+        }
+
+        let mut parts = self.host.clone().into_parts();
+        parts.path_and_query = Some(API_TOKEN.clone());
+        let uri = Uri::from_parts(parts)?;
+
+        let req = Request::put(uri)
+            .header("X-aws-ec2-metadata-token-ttl-seconds", "21600")
+            .body(Body::empty())?;
+
+        let res = self.client.request(req).compat().await?;
+
+        if res.status() != StatusCode::OK {
+            return Err(Ec2MetadataError::UnableToFetchToken.into());
+        }
+
+        let body = res.into_body().concat2().compat().await?;
+        let token = body.into_bytes();
+
+        let next_refresh = Instant::now() + Duration::from_secs(21600);
+        self.token = Some((token.clone(), next_refresh));
+
+        Ok(token)
+    }
+
+    pub async fn get_document(&mut self) -> Result<Option<DynamicIdentityDocument>, crate::Error> {
+        let token = self.get_token().await?;
+
+        let mut parts = self.host.clone().into_parts();
+        parts.path_and_query = Some(DYNAMIC_DOCUMENT.clone());
+        let uri = Uri::from_parts(parts)?;
+
+        let req = Request::get(uri)
+            .header(TOKEN_HEADER.clone(), token)
+            .body(Body::empty())?;
+
+        let res = self.client.request(req).compat().await?;
+
+        if res.status() != StatusCode::OK {
+            warn!(message="Dynamic request failed.", status = %res.status());
```

Comment:
Is there a better way to describe this than `Dynamic request`?

Reply:
ah there is a better name `instance identity document`, I will update that and related items.